### ISSUE TITLE
Performance regression on running `pnpm install`

### ISSUE THREAD

**Issue by bryhasagithub** [pnpm/pnpm#7297]
### Last pnpm version that worked

8.9.2

### pnpm version

8.10.2

### Code to reproduce the issue

Discord Discussion thread can be found [here](https://discord.com/channels/731599538665553971/1172647285507760128)

### Expected behavior

Before when I ran pnpm install on version 8.6.3
I would get this output

``` bash
➜  git:(dev) ✗ pnpm i
Scope: all 71 workspace projects
Lockfile is up to date, resolution step is skipped
Packages: +1
+
Progress: resolved 1, reused 0, downloaded 1, added 1, done
Done in 1.7s
```

### Actual behavior

``` bash
➜  git:(dev) ✗ pnpm i
Scope: all 71 workspace projects
Lockfile is up to date, resolution step is skipped
Packages: +3827
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Progress: resolved 1, reused 1, downloaded 0, added 1, done
Done in 5.6s
```

### Discord Discussion thread 
> The full conversation discord thread for where I discovered this to be a regression can be found [here](https://discord.com/channels/731599538665553971/1172647285507760128)

### Additional info
@zkochan replied to my thread on discord addressing that the issue appears to be a regression that may have been caused by this [change](https://github.com/pnpm/pnpm/pull/7214) that derived from this [issue](https://github.com/pnpm/pnpm/issues/5965). They confirmed that there is no existing github issue for this issue and so I've created a regression issue to track it's status.

### TL;DR:
- Running `pnpm install` on v8.6.3, the cached install time comes out to be roughly 1 or 2 seconds and the packages logged out that it analyzed is only +1. 
- Running `pnpm install` on v8.10.2, the cached install time comes out to be roughly 5- to 7 seconds and the packages logged out that it analyzed is all +3827.

### Screenshots:
**Before**
<img width="583" alt="before" src="https://github.com/pnpm/pnpm/assets/11837954/ae5a3a78-5fb0-41ba-b221-9610228ebc0b">
**After**
<img width="732" alt="after" src="https://github.com/pnpm/pnpm/assets/11837954/eedfd494-bcf4-4c02-8346-ad5ef7b80364">

**Big thank you to @zkochan for helping me debug this behavior I was having.** 

### Node.js version

16.20.0

### Operating System

macOS

**Comment by zkochan** [comment]
This PR has caused the issue: https://github.com/pnpm/pnpm/pull/7214
